### ISSUE TITLE
[Feature request] Densely crop patches through a whole image.

### ISSUE THREAD

**Issue by LambdaWill** [pytorch/pytorch#3387]
I am trying to transfer [Fast Patch-based Style Transfer of Arbitrary Style](https://arxiv.org/abs/1612.04337)(based on troch) to pytorch. Then I realize these for-loop is too slow, especially when
`img` is large.
```lua
function module._extract_patches(img, patch_size, stride)
    local nDim = 3
    assert(img:nDimension() == nDim, 'image must be of dimension 3.')
    local C, H, W = img:size(nDim-2), img:size(nDim-1), img:size(nDim)
    local nH = math.floor( (H - patch_size)/stride + 1)
    local nW = math.floor( (W - patch_size)/stride + 1)

    -- extract patches
    local patches = torch.Tensor(nH*nW, C, patch_size, patch_size):typeAs(img)
    for i=1,nH*nW do
        local h = math.floor((i-1)/nW)  -- zero-index
        local w = math.floor((i-1)%nW)  -- zero-index
        patches[i] = img[{{},
        {1 + h*stride, 1 + h*stride + patch_size-1},
        {1 + w*stride, 1 + w*stride + patch_size-1}
        }]
    end

    return patches
end
```
Can the operation be implemented in an efficient way?

**Comment by fmassa** [comment]
You can use the `unfold` function twice for it. See for example the snippet in https://github.com/pytorch/pytorch/pull/1523

**Comment by LambdaWill** [comment]
@fmassa Wow, I have managed to work it out according to your snippet. 
By the way, I wonder how to handle such situation.
```lua
function module._extract_patches(img, patch_size, stride, threshold)
    local nDim = 3
    assert(img:nDimension() == nDim, 'image must be of dimension 3.')
    local C, H, W = img:size(nDim-2), img:size(nDim-1), img:size(nDim)
    local nH = math.floor( (H - patch_size)/stride + 1)
    local nW = math.floor( (W - patch_size)/stride + 1)

    -- extract patches
    local patches = torch.Tensor(nH*nW, C, patch_size, patch_size):typeAs(img):zero()
    for i=1,nH*nW do
        local h = math.floor((i-1)/nW)  -- zero-index
        local w = math.floor((i-1)%nW)  -- zero-index

----------- Add a condition here ---------------------
        local tmp_img = img[{{},
        {1 + h*stride, 1 + h*stride + patch_size-1},
        {1 + w*stride, 1 + w*stride + patch_size-1}
        }]
        if torch.mean(tmp_img) > threshold then
               patches[i] = tmp_img
        end
    end

    return patches
end
```

**Comment by fmassa** [comment]
For your specific case, it seems that you can extract all the patches using unfold, call `.contiguous()`, and then zero out the indices that don't satisfy the condition.
Pseudo-code (untested):
```python
input_windows = input.unfold(2, kh, dh).unfold(3, kw, dw)
# view the windows as (kh * kw)
input_windows = input_windows.contiguous().view(*input_windows.size()[:-2], -1)
condition_windows = input_windows.mean(4) <= threshold
# mask the inputs you don't want
input_windows[condition_windows[..., None].expand_as(input_windows)] = 0
```

**Comment by FDSJK** [comment]
How can deal with the margin problem when I using unfold function for many times to extract 3d patches, because I found that some of pixels in margin will be ignored when the patch size can not be exactly divided.

**Comment by fmassa** [comment]
@FDSJK you might need to apply some padding on your input in order to achieve what you want.
For that, you can use `torch.nn.functional.unfold`, which supports a `padding` option. Alternatively, you can also manually pad your input.
